include/zephyr/net/wifi_credentials.h

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef WIFI_CREDENTIALS_H__
+#define WIFI_CREDENTIALS_H__
+
+#include <zephyr/types.h>
+#include <zephyr/net/wifi.h>
+#include <zephyr/kernel.h>
+
+/**
+ * @defgroup wifi_credentials Wi-Fi credentials library
+ * @ingroup networking
+ * @since 4.0
+ * @version 0.1.0
+ * @{
+ * @brief Library that provides a way to store and load Wi-Fi credentials.
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* this entry contains a BSSID */
+#define WIFI_CREDENTIALS_FLAG_BSSID        BIT(0)
+/* this entry is to be preferred over others */
+#define WIFI_CREDENTIALS_FLAG_FAVORITE     BIT(1)
+/* this entry can use the 2.4 GHz band */
+#define WIFI_CREDENTIALS_FLAG_2_4GHz       BIT(2)
+/* this entry can use the 5 GHz band */
+#define WIFI_CREDENTIALS_FLAG_5GHz         BIT(3)
+/* this entry requires management frame protection */
+#define WIFI_CREDENTIALS_FLAG_MFP_REQUIRED BIT(4)
+/* this entry disables management frame protection */
+#define WIFI_CREDENTIALS_FLAG_MFP_DISABLED BIT(5)
+
+#define WIFI_CREDENTIALS_MAX_PASSWORD_LEN                                                          \
+	MAX(WIFI_PSK_MAX_LEN, CONFIG_WIFI_CREDENTIALS_SAE_PASSWORD_LENGTH)
+
+/**
+ * @brief Wi-Fi credentials entry header
+ * @note Every settings entry starts with this header.
+ *       Depending on the `type` field, the header can be casted to a larger type.
+ *       In addition to SSID (usually a string) and BSSID (a MAC address),
+ *       a `flags` field can be used to control some detail settings.
+ *
+ */
+struct wifi_credentials_header {
+	enum wifi_security_type type;
+	char ssid[WIFI_SSID_MAX_LEN];
+	size_t ssid_len;
+	uint8_t bssid[WIFI_MAC_ADDR_LEN];
+	uint32_t flags;
+	uint8_t channel;
+	uint32_t timeout;
```

Comment:
This looks like it will leave a hole in memory. If you organize the fields differently we can save couple of bytes.
Suggesting something like this
```
	enum wifi_security_type type;
	char ssid[WIFI_SSID_MAX_LEN];
	size_t ssid_len;
	uint32_t flags;
	uint32_t timeout;
	uint8_t bssid[WIFI_MAC_ADDR_LEN];
	uint8_t channel;
```
Also the fields need to be documented, otherwise the generated documentation will look like crap.
Same comment about field documentation to other structs below.